internal/model.go

```diff
@@ -0,0 +1,229 @@
+ *
+ */
+
+package internal
+
+import (
+	"context"
+)
+
+const (
+
+	// plugin types
+	//TODO: where should this live?
+
+	PluginTypeKafka  = "kafka"  // generic kafka plugin
+	PluginTypeKDS    = "kds"    // generic kds plugin
+	PluginTypeSQS    = "sqs"    // generic sqs plugin
+	PluginTypeHTTP   = "http"   // generic http / webhook plugin
+	PluginTypeEars   = "ears"   // loopback plugin, may be useful for multi stage processing
+	PluginTypeGears  = "gears"  // deliver to kafka in masheens envelope
+	PluginTypeNull   = "null"   // black whole for events for testing
+	PluginTypeDebug  = "debug"  // debug plugin
+	PluginTypeFilter = "filter" // a filter plugin that filters or transforms and event
+
+	// filter types
+	//TODO: where should this live?
+
+	FilterTypeFilter      = "filter"
+	FilterTypeMatcher     = "match"
+	FilterTypeTransformer = "transform"
+	FilterTypeTTLer       = "ttl"
+	FilterTypeSplitter    = "split"
+	FilterTypePass        = "pass"
+	FilterTypeBlock       = "block"
+
+	// plugin modes
+
+	PluginModeInput  = "input"  // input plugin
+	PluginModeOutput = "output" // output plugin
+	PluginModeFilter = "filter" // filter plugin
+
+	// plugin states
+
+	PluginStateRunning      = "running"       // running -> shutting_down, error
+	PluginStateStopped      = "stopped"       // stopped -> ready
+	PluginStateError        = "error"         // error
+	PluginStateShuttingDown = "shutting_down" // shutting_down -> stopped
+	PluginStateReady        = "ready"         // * ready -> running
+
+	// delivery modes
+
+	DeliveryModeFireAndForget = "fire_and_forget"  // delivery not guranteed
+	DeliveryModeAtLeastOnce   = "at_least_once"    // duplicates allowed, message loss not allowed
+	DeliveryModeExactlyOnce   = "exactly_once"     // most precise delivery mode
+	DeliveryModeMostOfTheTime = "most_of_the_time" // realistic delivery mode
+	DeliveryModeNeverEver     = "never_ever"       // in case of message fatigue
+
+	// event encodings
+
+	EncodingJson = "json"
+)
+
+type (
+	// A FilterChain is a slice of filter plugins
+	FilterChain struct {
+		Filters []*FilterPlugin `json:"filters"` // optional list of filter plugins that will be applied in order to perform arbitrary filtering and transformation functions
+	}
+
+	// A RoutingTable is a slice of routing entries and reprrsents the EARS routing table
+	RoutingTable []*Route
+
+	// A RoutingTableIndex is a hashmap mapping a routing entry hash to a routing entry pointer
+	RoutingTableIndex map[string]*Route
+
+	// A PluginIndex is a hashmap mapping a plugin instance hash to a plugin instance
+	PluginIndex map[string]Pluginer
+)
+
+type (
+	Pluginer interface {
+		Hasher
+		Validater
+		Initializer
+		Closer
+		SyncDoer
+		AsyncDoer
+		GetConfig() *Plugin
+		GetInputChannel() chan *Event
+		GetOutputChannel() chan *Event
+		GetRouteCount() int
+		GetEventCount() int
+	}
+
+	// A Hasher can provide a unique deterministic hash string based on its configuration parameters
```

Comment:
This comment is about Hasher. Might want to update it to talk about Closer

Reply:
ok